receiver/vcenterreceiver/factory.go

```diff
@@ -0,0 +1,117 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vcenterreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver"
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configtls"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver/internal/metadata"
+)
+
+const (
+	typeStr = "vcenter"
+)
+
+type vcenterReceiverFactory struct {
+	receivers map[*Config]*vcenterReceiver
+}
+
+// NewFactory returns the receiver factory for the vcenterreceiver
+func NewFactory() component.ReceiverFactory {
+	f := &vcenterReceiverFactory{
+		receivers: make(map[*Config]*vcenterReceiver),
+	}
+	return component.NewReceiverFactory(
+		typeStr,
+		createDefaultConfig,
+		component.WithMetricsReceiver(f.createMetricsReceiver),
+	)
+}
+
+func createDefaultConfig() config.Receiver {
+	return &Config{
+		MetricsConfig: &MetricsConfig{
+			ScraperControllerSettings: scraperhelper.ScraperControllerSettings{
+				ReceiverSettings:   config.NewReceiverSettings(config.NewComponentID(typeStr)),
+				CollectionInterval: 5 * time.Minute,
+			},
+			TLSClientSetting: configtls.TLSClientSetting{},
+			Metrics:          metadata.DefaultMetricsSettings(),
+			Endpoint:         "",
+			Username:         "",
+			Password:         "",
+		},
+	}
+}
+
+func (f *vcenterReceiverFactory) ensureReceiver(params component.ReceiverCreateSettings, config config.Receiver) *vcenterReceiver {
```

Comment:
It seems a little backwards to me that we are checking these things after doing all the other initialization. Is there a reason to do it that way?

Reply:
The purpose of this function is to enforce the singleton pattern, otherwise the collector gets different pointers for the receiver for logs and metrics and then will refuse to start. This shouldn't be an error state we always want to have 1 instance of the receiver be the same pointer. 

Does that answer your question? Or do you think we should have a different name for ensuring that the singleton is enforced?

Reply:
What I'm confused about is why it has to happen after creating the scraper and doing all the other checking in the calling function. Why not do this first and then enrich the receiver?

Reply:
Just pushed up a change that does this 👍